{
 "metadata": {
  "name": "Session11"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Session 11 : Parsing BLAST results\n",
      "\n",
      "## Learning Objectives\n",
      "\n",
      "* Be able to parse a BLAST results file\n",
      "\n",
      "### Overview\n",
      "\n",
      "If you ran the following command from exercise 4 in Session 10 \n",
      "\n",
      "<pre>\n",
      "blastall -p blastp -i NC_010001.faa -d all.faa -e .001 -o NC_010001_results.txt\n",
      "</pre>\n",
      "\n",
      "You would get the following result for the first BLAST query\n",
      "\n",
      "<pre>\n",
      "BLASTP 2.2.26 [Sep-21-2011]\n",
      "\n",
      "\n",
      "Reference: Altschul, Stephen F., Thomas L. Madden, Alejandro A. Schaffer, \n",
      "Jinghui Zhang, Zheng Zhang, Webb Miller, and David J. Lipman (1997), \n",
      "\"Gapped BLAST and PSI-BLAST: a new generation of protein database search\n",
      "programs\",  Nucleic Acids Res. 25:3389-3402.\n",
      "\n",
      "Reference for compositional score matrix adjustment: Altschul, Stephen F., \n",
      "John C. Wootton, E. Michael Gertz, Richa Agarwala, Aleksandr Morgulis,\n",
      "Alejandro A. Schaffer, and Yi-Kuo Yu (2005) \"Protein database searches\n",
      "using compositionally adjusted substitution matrices\", FEBS J. 272:5101-5109.\n",
      "\n",
      "Query= gi|160878163|ref|YP_001557131.1| chromosomal replication\n",
      "initiator protein DnaA [Clostridium phytofermentans ISDg]\n",
      "         (453 letters)\n",
      "\n",
      "Database: all.faa \n",
      "           9864 sequences; 3,021,451 total letters\n",
      "\n",
      "Searching..................................................done\n",
      "\n",
      "\n",
      "\n",
      "                                                                 Score    E\n",
      "Sequences producing significant alignments:                      (bits) Value\n",
      "\n",
      "ref|YP_002935946.1| chromosomal replication initiation protein [...   484   e-170\n",
      "ref|YP_001385963.1| chromosomal replication initiation protein [...   460   e-161\n",
      "ref|YP_005686552.1| chromosomal replication initiator protein Dn...   431   e-149\n",
      "ref|YP_001389328.1| DNA replication protein DnaC [Clostridium bo...    42   8e-05\n",
      "ref|YP_002936373.1| hypothetical protein EUBREC_0448 [Eubacteriu...    42   2e-04\n",
      "\n",
      ">ref|YP_002935946.1| chromosomal replication initiation protein [Eubacterium rectale\n",
      "           ATCC 33656]\n",
      "          Length = 453\n",
      "\n",
      " Score =  484 bits (1247), Expect = e-170,   Method: Compositional matrix adjust.\n",
      " Identities = 242/455 (53%), Positives = 320/455 (70%), Gaps = 14/455 (3%)\n",
      "\n",
      "Query: 5   IQEKWNEILEFLKIEYNVTEVSYKTWLLPLKVYDVKDNVIKLSVDDTKIGANSLDFIKNK 64\n",
      "           + EKWNEIL+ +K E+ +T+VS+ TW+ PL+V+ +  N + + V   ++G   L +I  K\n",
      "Sbjct: 4   VMEKWNEILQMVKEEHGLTDVSFNTWIKPLEVFAIDGNTLYILVPSEQMG---LSYINKK 60\n",
      "\n",
      "Query: 65  YSQFLKTAIAEVINQDFEIEFVLLSQTKA-------EEKVQTQAPNKIKNESLSYLNPRY 117\n",
      "           Y   LK AI EV   +++I+F+L  Q ++         +++ Q P K  +   S LNP Y\n",
      "Sbjct: 61  YYLPLKVAIGEVTGIEYDIQFILPDQAQSLKFKNNDTPQLEQQTPVKGDH---SNLNPNY 117\n",
      "\n",
      "Query: 118 TFDTFVVGXXXXXXXXXXXXXXXXXXXIYNPLFIYGGVGLGKTHLMHSIAHYILEQNPNS 177\n",
      "           TFDTFVVG                    YNPL+IYGG GLGKTHLMHSI H+I++QNP++\n",
      "Sbjct: 118 TFDTFVVGNNNRFAHSASLAVAESPGEAYNPLYIYGGPGLGKTHLMHSIGHFIIDQNPDT 177\n",
      "\n",
      "Query: 178 KVLYVTSEKFTNELIESIRNADTTP-TEFREKYRNIDVLLIDDIQFIIGKERTQEEFFHT 236\n",
      "           KVLYVTSE+FTNE+IESIRN +++  T+FR+KYR +DVL+IDDIQFIIGKE TQEEFFHT\n",
      "Sbjct: 178 KVLYVTSEEFTNEVIESIRNGNSSAMTKFRDKYRTVDVLMIDDIQFIIGKESTQEEFFHT 237\n",
      "\n",
      "Query: 237 FNTLHESKKQIIISSDKPPKDILTLEERLRSRFEWGLTVDIQSPDYETRMAILKKKEELD 296\n",
      "           FNTL    KQII++SDKPPK++ TLEER+RSRFEWGL  DI +PDYETRMAIL+KK E D\n",
      "Sbjct: 238 FNTLQTQGKQIILTSDKPPKEMETLEERIRSRFEWGLMADIGTPDYETRMAILRKKAETD 297\n",
      "\n",
      "Query: 297 CLTIDDEVMKYIASNIKSNIRELEGALTKIVALSRLKKKEVDVILAEEALKDLISPDNKK 356\n",
      "              IDD+++ YIASNI SNIRELEGAL K++A   L+   + + +AE  L ++I+PD  +\n",
      "Sbjct: 298 NFDIDDDILNYIASNISSNIRELEGALNKLLAFHNLEHTHITMDIAERELSNIITPDKPR 357\n",
      "\n",
      "Query: 357 TVTLDLIIEVVSEHFTTSTSEIYSDNRSRNIAYPRQIAMYLCRKLTSLSLTDIGKMMGNR 416\n",
      "            +T  LIIEVVSEHF  S  ++ S  RS  IA PRQIAMYLC+ +TS SL  IG+++G R\n",
      "Sbjct: 358 EITAQLIIEVVSEHFHISVDQMISKTRSSEIARPRQIAMYLCKTMTSDSLDVIGQLLGGR 417\n",
      "\n",
      "Query: 417 DHSTVLHGCNKVEKDIKKDPSFQNTIDVLIKKINP 451\n",
      "           DHST++HG  KV KD +++ S +  I+ + KKINP\n",
      "Sbjct: 418 DHSTIIHGIKKVTKDYEENDSTRTLIETIKKKINP 452\n",
      "\n",
      "\n",
      ">ref|YP_001385963.1| chromosomal replication initiation protein [Clostridium botulinum A\n",
      "           str. Hall]\n",
      "          Length = 448\n",
      "\n",
      " Score =  460 bits (1184), Expect = e-161,   Method: Compositional matrix adjust.\n",
      " Identities = 236/456 (51%), Positives = 313/456 (68%), Gaps = 16/456 (3%)\n",
      "\n",
      "Query: 1   MKSLIQEKWNEILEFLKIEYNVTEVSYKTWLLPLKVYDVKDNVIKLSV--DDTKIGANSL 58\n",
      "           M + + E W + +  +K E  +TEVS+ TW+  +    +++N +KL+V  D TK      \n",
      "Sbjct: 1   MNTHLTETWEKAINIIKGE--LTEVSFNTWIKSINPISLENNSLKLAVPNDFTK------ 52\n",
      "\n",
      "Query: 59  DFIKNKYSQFLKTAIAEVINQDFEIEFVLLSQTKAEEKVQTQAPNK----IKNESLSYLN 114\n",
      "             ++++Y   +  A+  + ++ + I+F++ ++ K EE  +     K    + +E  + LN\n",
      "Sbjct: 53  GILESRYKDLIVNALKLLTSKKYNIDFIVTTEEKIEENQKNHNNEKSNIVVNDEMSTMLN 112\n",
      "\n",
      "Query: 115 PRYTFDTFVVGXXXXXXXXXXXXXXXXXXXIYNPLFIYGGVGLGKTHLMHSIAHYILEQN 174\n",
      "           P+YTFD+FV+G                    YNPLFIYGGVGLGKTHLMH+I HYIL  N\n",
      "Sbjct: 113 PKYTFDSFVIGNSNRFAHAASLAVAESPAKAYNPLFIYGGVGLGKTHLMHAIGHYILHNN 172\n",
      "\n",
      "Query: 175 PNSKVLYVTSEKFTNELIESIRNADTTPTEFREKYRNIDVLLIDDIQFIIGKERTQEEFF 234\n",
      "           P S+V+YV+SEKFTNELI SI+  D    EFR KYRNID+LL+DDIQFI GKERTQEEFF\n",
      "Sbjct: 173 PKSQVVYVSSEKFTNELINSIK--DDKNVEFRNKYRNIDILLVDDIQFIAGKERTQEEFF 230\n",
      "\n",
      "Query: 235 HTFNTLHESKKQIIISSDKPPKDILTLEERLRSRFEWGLTVDIQSPDYETRMAILKKKEE 294\n",
      "           HTFN L+E+ KQIIISSD+PPK+I TLE+RLRSRFEWGL  DIQ+PD+ETRMAILKKK +\n",
      "Sbjct: 231 HTFNALYEANKQIIISSDRPPKEIPTLEDRLRSRFEWGLIADIQAPDFETRMAILKKKAD 290\n",
      "\n",
      "Query: 295 LDCLTIDDEVMKYIASNIKSNIRELEGALTKIVALSRLKKKEVDVILAEEALKDLISPDN 354\n",
      "           ++ L I +EVM YIA+ IKSNIRELEGAL +IVA S L  KE+ + LA EALKD+IS   \n",
      "Sbjct: 291 VENLNIPNEVMVYIATKIKSNIRELEGALIRIVAFSSLTNKEISIDLASEALKDIISSKQ 350\n",
      "\n",
      "Query: 355 KKTVTLDLIIEVVSEHFTTSTSEIYSDNRSRNIAYPRQIAMYLCRKLTSLSLTDIGKMMG 414\n",
      "            + VT+D+I EVV+ ++     ++ S  R+RNIA+PRQIAMYL RKLT +SL  IG+  G\n",
      "Sbjct: 351 TRQVTIDIIQEVVANYYNLKIEDLKSARRTRNIAFPRQIAMYLSRKLTDMSLPKIGEEFG 410\n",
      "\n",
      "Query: 415 NRDHSTVLHGCNKVEKDIKKDPSFQNTIDVLIKKIN 450\n",
      "            RDH+TV+H   K+  ++KKD S QN I  L K+IN\n",
      "Sbjct: 411 GRDHTTVIHAYEKISNNLKKDESLQNAIKELNKRIN 446\n",
      "\n",
      "\n",
      ">ref|YP_005686552.1| chromosomal replication initiator protein DnaA [Clostridium\n",
      "           thermocellum DSM 1313]\n",
      "          Length = 443\n",
      "\n",
      " Score =  431 bits (1108), Expect = e-149,   Method: Compositional matrix adjust.\n",
      " Identities = 219/435 (50%), Positives = 294/435 (67%), Gaps = 9/435 (2%)\n",
      "\n",
      "Query: 1   MKSLIQEKWNEILEFLKIEYNVTEVSYKTWLLPLKVYDVKDNVIKLSVDDTKIGANSLDF 60\n",
      "           M + + E W + L  LK E  +TE+S+ TW+  +    +  N I L+V        +   \n",
      "Sbjct: 1   MNTQLNEIWQKTLGLLKNE--LTEISFNTWIKTIDPLSLTGNTINLAVP----AEFNKGI 54\n",
      "\n",
      "Query: 61  IKNKYSQFLKTAIAEVINQDFEIEFVLLSQTKAEEKVQTQAPNKIKNESLSYLNPRYTFD 120\n",
      "           ++++Y   +K AI +V  +++EI F++ SQ    +  +       ++  LS LNP+YTFD\n",
      "Sbjct: 55  LESRYQTLIKNAIKQVTFKEYEIAFIVPSQENLNKLTKQTESAGNEDSPLSVLNPKYTFD 114\n",
      "\n",
      "Query: 121 TFVVGXXXXXXXXXXXXXXXXXXXIYNPLFIYGGVGLGKTHLMHSIAHYILEQNPNSKVL 180\n",
      "           TFV+G                    YNPLFIYGGVGLGKTHLMH+I HYILEQN + +VL\n",
      "Sbjct: 115 TFVIGNSNRFAHAAALAVAEAPGKAYNPLFIYGGVGLGKTHLMHAIGHYILEQNSSQRVL 174\n",
      "\n",
      "Query: 181 YVTSEKFTNELIESIRNADTTPTEFREKYRNIDVLLIDDIQFIIGKERTQEEFFHTFNTL 240\n",
      "           YV+SEKFTNELI +I+  D    EFR KYRNIDVLLIDDIQFI GKERT+EEFFHTFN L\n",
      "Sbjct: 175 YVSSEKFTNELINAIK--DNRNEEFRSKYRNIDVLLIDDIQFIAGKERTEEEFFHTFNAL 232\n",
      "\n",
      "Query: 241 HESKKQIIISSDKPPKDILTLEERLRSRFEWGLTVDIQSPDYETRMAILKKKEELDCLTI 300\n",
      "           +E+ KQII+SSDKPPK+I +LE+RLRSRFEWGL  D+Q+PD ETR+AIL+KK +L+ LT+\n",
      "Sbjct: 233 YEANKQIILSSDKPPKEI-SLEDRLRSRFEWGLIADMQAPDLETRIAILRKKAQLENLTV 291\n",
      "\n",
      "Query: 301 DDEVMKYIASNIKSNIRELEGALTKIVALSRLKKKEVDVILAEEALKDLISPDNKKTVTL 360\n",
      "            +EV+ +IA  I SNIRELEGAL +++A S L + E+ V LA EALKD++S +  K +  \n",
      "Sbjct: 292 PNEVIVFIADKIASNIRELEGALNRVIAYSSLTENEITVELASEALKDILSANKAKVLNC 351\n",
      "\n",
      "Query: 361 DLIIEVVSEHFTTSTSEIYSDNRSRNIAYPRQIAMYLCRKLTSLSLTDIGKMMGNRDHST 420\n",
      "             I E V+ +F     E  S  R+R+IA+PRQIAMYLCR+LT +SL  IG+  G RDH+T\n",
      "Sbjct: 352 TTIQEAVARYFDIRPEEFKSKKRTRDIAFPRQIAMYLCRELTEMSLPKIGEEFGGRDHTT 411\n",
      "\n",
      "Query: 421 VLHGCNKVEKDIKKD 435\n",
      "           V+H C K+ ++I+ +\n",
      "Sbjct: 412 VIHACEKISEEIESN 426\n",
      "\n",
      "\n",
      ">ref|YP_001389328.1| DNA replication protein DnaC [Clostridium botulinum A str. Hall]\n",
      "          Length = 325\n",
      "\n",
      " Score = 42.4 bits (98), Expect = 8e-05,   Method: Compositional matrix adjust.\n",
      " Identities = 42/131 (32%), Positives = 71/131 (54%), Gaps = 9/131 (6%)\n",
      "\n",
      "Query: 147 NPLFIYGGVGLGKTHLMHSIAHYILEQNPNSKVLYVTSEKFTNELIESIRNADTTPTEFR 206\n",
      "           N +FI G  G GKT L + IA  +L++   + V+Y T++    ELI+++R+   +  +  \n",
      "Sbjct: 185 NFMFI-GTPGTGKTFLSNCIAKELLDRG--NFVVYRTAD----ELIQNLRSIKFSNDKHL 237\n",
      "\n",
      "Query: 207 EKYR-NIDVLLIDDIQFIIGKERTQEEFFHTFNTLHESKKQIIISSDKPPKDIL-TLEER 264\n",
      "           E    N D+L+IDD+      E ++ E F+  N     +K++IISS+   + IL +  ER\n",
      "Sbjct: 238 EDILINCDLLIIDDLGTESINEFSKVELFNFINKKLLMRKKMIISSNYSIETILKSYSER 297\n",
      "\n",
      "Query: 265 LRSRFEWGLTV 275\n",
      "           + SR     T+\n",
      "Sbjct: 298 ISSRLLGNFTL 308\n",
      "\n",
      "\n",
      ">ref|YP_002936373.1| hypothetical protein EUBREC_0448 [Eubacterium rectale ATCC 33656]\n",
      "          Length = 327\n",
      "\n",
      " Score = 41.6 bits (96), Expect = 2e-04,   Method: Compositional matrix adjust.\n",
      " Identities = 33/121 (27%), Positives = 56/121 (46%), Gaps = 3/121 (2%)\n",
      "\n",
      "Query: 149 LFIYGGVGLGKTHLMHSIAHYILEQNPNSKVLYVTSEKFTNELIESIRNADTTPTEFREK 208\n",
      "           +F YG  G+GKT L + IA  +L+      V+Y T+ +  + L + +   D       + \n",
      "Sbjct: 186 IFFYGNTGVGKTFLSNCIAKELLD--AGYSVIYFTAFQLFDILSKGVFEKDADAIAAHQN 243\n",
      "\n",
      "Query: 209 YRNIDVLLIDDIQFIIGKERTQEEFFHTFNTLHESKKQIIISSDKPPKDILTL-EERLRS 267\n",
      "             + D+L+IDD+   +    T  + F   N     +K  IIS++   + I  +  ER  S\n",
      "Sbjct: 244 IFDCDLLIIDDLGTELSNSFTTSQLFLCVNERILRQKSTIISTNLNLEQIADIYSERTLS 303\n",
      "\n",
      "Query: 268 R 268\n",
      "           R\n",
      "Sbjct: 304 R 304\n",
      "</pre>\n",
      "\n",
      "### Parsing BLAST results\n",
      "\n",
      "In computer programming parsing is the process of analysing a file according to a set of rules that are specified in the code.  To parse a file you need to know something about the structure of the text in the file.  For example in the above BLAST results example each new result begins with the line BLASTP 2.2.26 [Sep-21-2011].  Thus you could determine the number of separate results by counting the times this line appears.  The fasta header from each query sequence also appears in each result starting with the line Query=.  "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/env python\n",
      "\n",
      "# Example 11.1 \n",
      "# Name: blast_results_count.py\n",
      "# Description:  This program counts the number of separate blast results in a file\n",
      "# Usage: python blast_results_count.py filename \n",
      "# Example: python blast_results_count.py NC_010001_results.txt\n",
      "\n",
      "import sys\n",
      "\n",
      "# Add a check to see if there is the right number of arguments \n",
      "# Exit if there is not the correct number\n",
      "if len(sys.argv) != 2 :\n",
      "\tprint \"This program takes a filename as an argumnet\"\n",
      "\tsys.exit(2)\n",
      "    \n",
      "# Open the file\n",
      "blast_file = open(sys.argv[1], 'r')\n",
      "\n",
      "# read the file into the list blast_results\n",
      "blast_results = blast_file.readlines()\n",
      "\n",
      "# Initialize counts\n",
      "count = 0\n",
      "\n",
      "for line in blast_results:\n",
      "\tif line.startswith(\"BLASTP 2.2.26\"): \n",
      "\t\tcount += 1\n",
      "print \"There are %s separate query results in this file.\" % (count)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "SystemExit",
       "evalue": "2",
       "output_type": "pyerr",
       "traceback": [
        "An exception has occurred, use %tb to see the full traceback.\n",
        "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "This program takes a filename as an argumnet\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "To exit: use 'exit', 'quit', or Ctrl-D."
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Setting flags\n",
      "\n",
      "One way to capture the just the first blast hit would be to use flags to determine if this is a new result and change the flag once the top hit has been read. "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/env python\n",
      "\n",
      "# Example 11.2 \n",
      "# Name: flag_top_hit.py\n",
      "# Description:  This program captures the sequenc alignment for the top blast hit\n",
      "# Usage: python flag_top_hit.py filename \n",
      "# Example: python flag_top_hit.py NC_010001_5_results2.txt\n",
      "\n",
      "import sys\n",
      "\n",
      "# Add a check to see if there is the right number of arguments \n",
      "# Exit if there is not the correct number\n",
      "if len(sys.argv) != 2 :\n",
      "\tprint \"This program takes a filename as an argumnet\"\n",
      "\tsys.exit(2)\n",
      "    \n",
      "# Open the file\n",
      "blast_file = open(sys.argv[1], 'r')\n",
      "\n",
      "# read the file into the list blast_results\n",
      "blast_results = blast_file.readlines()\n",
      "\n",
      "new_blast_flag = 'FALSE'\n",
      "\n",
      "for line in blast_results:\n",
      "\tif line.startswith(\"Sequences producing significant alignments\"):\n",
      "\t\tnew_blast_flag = 'TRUE'\n",
      "\tif (new_blast_flag == 'TRUE') and line.startswith(\"ref|\"): \n",
      "\t\tprint line\n",
      "\t\tnew_blast_flag = 'FALSE'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Flags are also set using 1 and 0 instead of TRUE and FALSE. Below is an example that captures all the top BLAST hits."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/env python\n",
      "\n",
      "# Example 11.2 \n",
      "# Name: flag_hit_list.py\n",
      "# Description:  This program captures the sequenc alignment for the top blast hit\n",
      "# Usage: python flag_hit_list.py filename \n",
      "# Example: python flag_hit_list.py NC_010001_5_results2.txt\n",
      "\n",
      "import sys\n",
      "\n",
      "# Add a check to see if there is the right number of arguments \n",
      "# Exit if there is not the correct number\n",
      "if len(sys.argv) != 2 :\n",
      "\tprint \"This program takes a filename as an argumnet\"\n",
      "\tsys.exit(2)\n",
      "    \n",
      "# Open the file\n",
      "blast_file = open(sys.argv[1], 'r')\n",
      "\n",
      "# read the file into the list blast_results\n",
      "blast_results = blast_file.readlines()\n",
      "\n",
      "new_blast_flag = 'FALSE'\n",
      "\n",
      "for line in blast_results:\n",
      "\tif line.startswith(\"Sequences producing significant alignments\"):\n",
      "\t\tnew_blast_flag = 'TRUE'\n",
      "\t\tprint \"NEW BLAST RESULT\"\n",
      "\tif (new_blast_flag == 'TRUE') and line.startswith(\"ref|\"): \n",
      "\t\tprint line\n",
      "\tif (new_blast_flag == 'TRUE') and line.startswith(\">\"):  \n",
      "\t\tnew_blast_flag = 'FALSE'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Excercises\n",
      "\n",
      "Using the long format results of searching the C. phytofermentans proteome against the 3 other Clostridia genomes.\n",
      "\n",
      "A.  Write a program that extracts the query fasta header information below from all Query sequences (~4000)\n",
      "\n",
      "<pre>\n",
      "Query= gi|160878163|ref|YP_001557131.1| chromosomal replication\n",
      "initiator protein DnaA [Clostridium phytofermentans ISDg]\n",
      "         (453 letters)\n",
      "</pre>\n",
      "\n",
      "B.  Write a program that extracts the result fasta header information and the results for each top hit\n",
      "\n",
      "<pre>\n",
      ">ref|YP_002935946.1| chromosomal replication initiation protein [Eubacterium rectale\n",
      "           ATCC 33656]\n",
      "          Length = 453\n",
      "\n",
      " Score =  484 bits (1247), Expect = e-170,   Method: Compositional matrix adjust.\n",
      " Identities = 242/455 (53%), Positives = 320/455 (70%), Gaps = 14/455 (3%)\n",
      "</pre>\n",
      "C.  Write a program using the above 2 programs that gives as output a nice CSV formatted table of the results.  Using the first query as an example the results might look like\n",
      "\n",
      "<pre>\n",
      "query ref, query function, hit ref, hit function, hit organism, Score, Expect\n",
      "YP_001557131.1, chromosomal replication initiator protein DnaA, YP_002935946.1, Eubacterium rectale\n",
      "           ATCC 33656, 1247, e-170\n",
      "</pre>\n",
      "D.  Write a program using the above table as an input, that determines how many times each of the 3 organisms in the blast database come up as the best hit and include instances where no hit is above e = 0.001.\n",
      "\n",
      "BONUS.  Write a program that makes a bar chart from the best hit results in D."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "* Previous - <a href=\"http://nbviewer.ipython.org/urls/raw.github.com/jeffreyblanchard/AdvGen/master/Session10.ipynb\">Session 10 : BLAST</a> "
     ]
    }
   ],
   "metadata": {}
  }
 ]
}