{
 "metadata": {
  "name": "Session8"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Session 8 : Sequence Alignment and the Dynamic Programming Algorithm\n",
      "\n",
      "### Background\n",
      "\n",
      "There are three principal methods of pair-wise sequence alignment\n",
      "\n",
      "* Dot matrix analysis.\n",
      "* The dynamic programming (or DP) algorithm.\n",
      "\n",
      "### Dot matrix analysis\n",
      "\n",
      "The dot matrix methods displays graphically any possible sequence alignments as diagonals on the matrix.   Insertions, deletions, direct repeats and inverted repeats can be visually detected on a dot plot.   Dot plot can also detect regions of RNA that are self-complementary and thus might form secondary structure.\n",
      "\n",
      "\n",
      "<img src=\"https://bcrc.bio.umass.edu/courses/spring2012/micbio/micbio660/sites/default/files/dotplot.gif\" width=\"500\" alt=\"Dot plot\" /><br>\n",
      "\n",
      "Ref:<a href=\"http://mummer.sourceforge.net/manual/\"></a>\n",
      "\n",
      "This dot plot represents an alignment of two different strains of Helicobacter pylori. Forward matches are shown in red, while reverse matches are shown in green. This alignment clearly shows a major inversion event centered around the origin of replication.\n",
      "\n",
      "### The dynamic programming algorithm\n",
      "\n",
      "\"Dynamic programming was formalized by mathematician Richard Bellman, who was working at RNAD corporation on optimal decision processes.  He wanted to concoct an impressive name that would shield his work from US Secretary of Defense Charles Wilson, a man know to be hostile to mathematics research.   His work involved times series and planning - thus 'dynamic' and 'programming' (note, nothing particularly to do with computer programming).  Bellman especially liked 'dynamic' because \"it's impossible to use the work dynamic in the pejorative sense\"; he figured dynamic programming was \"something not even a Congressman could object to\" (Bellman 1984).\" (Eddy 2004)\n",
      "\n",
      "The dynamic programming algorithm is at the heart of many bioinformatics programs including BLAST (sequence searching), FASTA (sequence searching), CLUSTALW (multiple sequence alignmnet), HMMER (profile searching), GENSCAN (gene finding), MFOLD (RNA folding), and PHYLIP (phylogenetic analysis).   Dynamic programing was first used for global alignment of protein sequences by Needleman and Wunnsch (1970) and to find local alignments by Smith and Waterson (1981).\n",
      "\n",
      "<pre>\n",
      "  MLIGKVIGSVVSTRKN-LNVGKFMIVEPLK\n",
      "  |     |  |  ||   || ||||     |        Global Alignment\n",
      "  MVAARLI-VWANTRKTSLNLGKFMLAEIIK\n",
      "  \n",
      "              TRKN-LNVGKFM\n",
      "              ||   || ||||               Local Alignment\n",
      "              TRKTDLNLGKFML\n",
      "</pre>\n",
      "\n",
      "\n",
      "### Subsitution penalties differ for particular nucleotides or aminoacids< \n",
      "\n",
      "Amino acids are put together into proteins based on their three nucleotide codons, and most mutational events usually only change one nucleotide at time.  For example, starting with the Alanine codon GCU and looking at all possible changes on a http://en.wikipedia.org/wiki/Genetic_code#RNA_codon_table \n",
      "\n",
      "    GCU -> UCU = Alanine (A) -> Serine (S)\n",
      "    GCU -> ACU = Alanine (A) -> Threonine (T)\n",
      "    GCU -> CCU = Alanine (A) -> Proline (P)\n",
      "    GCU -> GUU = Alanine (A) -> Valine (V)\n",
      "    GCU -> GAU = Alanine (A) -> Aspartic Acid (D)\n",
      "    GCU -> GGU = Alanine (A) -> Glycine (G)\n",
      "    GCU -> GCA = Alanine (A) -> Alanine (A)\n",
      "    GCU -> GCC = Alanine (A) -> Alanine (A)\n",
      "    GCU -> GCG = Alanine (A) -> Alanine (A)\n",
      "\n",
      "Thus, from the Alanine codon GCU with a single substitution we can only get to 6 of the 19 other amino acids and 33% of the changes are going to result in keeping Alanine at the site in the protein, the other six amino acids would result 11% of the time.  Even if we look at all codons for Alanine, there are still 12 amino acids that can not be made by single substitution.  Thus, the frequency of changing Alanine to another amino acid by a single nucleotide change would be \n",
      "\n",
      "<pre>\n",
      "       A   R   N   D   C   Q   E   G   H   I   L   K   M   F   P   S   T   W   Y   V\n",
      "    A .33  0   0  .06  0   0  .06 .11  0   0   0   0   0   0  .11 .11 .11  0   0  .11\n",
      "</pre>\n",
      "\n",
      "### Origin of protein substitution matrices and bioinformatics\n",
      " \n",
      "Protein substitution matrices were first developed by Margaret Dayhoff, one of the founders of the field of bioinformatics.   She was particularly interested in the possibility of deducing the evolutionary relationships of organisms from protein sequences. Toward these ends she collected all the known protein sequences and, as a service to the scientific community, made them available to others in 1965 in a small book, the first Atlas of Protein Sequence and Structure.   \n",
      "\n",
      "Ref: http://en.wikipedia.org/wiki/Margaret_Oakley_Dayhoff\n",
      "\n",
      "From these sequences, she and her coworkers developed a model of protein evolution which resulted in the development of a set of widely used substitution matrices. These are frequently called Dayhoff or PAM (Percent Accepted Mutation) matrices and similar matrices are used in many areas of bioinformatics including: sequence similarity searches (e.g. BLAST), multiple sequence alignment tools (e.g ClustalW), phylogenetics and identifying functional regions of proteins.\n",
      "\n",
      "### Natural selection governs which amino acid changes are observed\n",
      "\n",
      "Just like mutation mucks with the frequency matrix so does Natural Selection.  Changes that occur between amino acids that have different biochemical properties are likely to affect the function of the protein.  Therefore, a substitution is more likely to occur between amino acids with similar biochemical properties.  In the above example with the Alanine codon a substitution that yields an amino acid change would result in mostly changes to other neutral amino acids, whereas the frequency of change to Aspartic Acid would probably be much lower. Amazingly, the genetic code has evolved to minimize changes between amino acids with different biochemical properties.\n",
      "\n",
      "<pre>\n",
      "Amino Acid          Side chain polarity Side chain acidity or basicity\n",
      " Alanine              nonpolar            neutral\n",
      " Arginine             polar                  strongly basic\n",
      " Asparagine           polar                  neutral\n",
      " Aspartic acid        polar                  acidic\n",
      " Cysteine             polar                  neutral\n",
      " Glutamic acid        polar                  acidic\n",
      " Glutamine            polar                  neutral\n",
      " Glycine              nonpolar               neutral\n",
      " Histidine            polar                  weakly basic\n",
      " Isoleucine           nonpolar               neutral\n",
      " Leucine              nonpolar               neutral\n",
      " Lysine               polar                  basic\n",
      " Methionine           nonpolar               neutral\n",
      " Phenylalanine        nonpolar               neutral\n",
      " Proline              nonpolar               neutral\n",
      " Serine               polar                  neutral\n",
      " Threonine            polar                  neutral\n",
      " Tryptophan           nonpolar               neutral\n",
      " Tyrosine             polar                  neutral\n",
      " Valine               nonpolar               neutral\n",
      "</pre>\n",
      "\n",
      "### Accounting for multiple substitutions\n",
      "\n",
      "As time goes on and sequence divergence gets larger it gets harder to account for multiple substitutions at the same amino acid position.  PAM protein matrices are based on global alignments of closely related proteins.  However, sequence changes over long evolutionary time scales are not well approximated by compounding small changes that occur over short time scales.  For comparing more distantly related sequence other types of matrices are used.  One of the most common is the The BLOSUM (BLOck SUbstitution Matrix) series of matrices created by Steve Henikoff and colleagues.  Both the PAM and BLOSUM are not expressed as transformation frequencies but the probabilities of transformation are expressed by log-odds scores as shown below for a BLOSUM62 matrix.\n",
      "  \n",
      "\n",
      "<img src=\"https://bcrc.bio.umass.edu/courses/spring2012/micbio/micbio660/sites/default/files/400px-blosum62.gif\" width=\"700\" alt=\"BLOSUM62\"\n",
      "\n",
      "\n",
      "Ref: http://en.wikipedia.org/wiki/Substitution_matrix\n",
      "\n",
      "\n",
      "### Dynamic programming examples\n",
      "\n",
      "We will go over the following dynamic programming examples in class\n",
      "\n",
      " * <a href=\"http://www.avatar.se/molbioinfo2001/dynprog/dynamic.html\"> Eric Rouchka's simple DP tutorial</a>\n",
      " * <a href=\"http://www.avatar.se/molbioinfo2001/dynprog/adv_dynamic.html\"> Eric Rouchka's advanced DP tutorial</a>\n",
      "\n",
      "Below is an example of the Dyanmic Programming Algorithm in Python implemented using a BLOSUM matrix.\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/env python\n",
      "\n",
      "# Example 8.1 \n",
      "# Name: Needle-Wunsch.py\n",
      "# Description: A program to demonstrate how Needle-Wunsch alignment algorithm\n",
      "# Adapted from Greg Caporaso - http://nbviewer.ipython.org/4657175/Lecture10.ipynb\n",
      "\n",
      "from __future__ import division\n",
      "from random import choice\n",
      "\n",
      "# Define a substitution matrix and a scorer function to look values up in that substitution matrix.\n",
      "\n",
      "blosum50 = {'A': {'A': 5.0,   'C': -1.0,   'D': -2.0,   'E': -1.0,   'F': -3.0,   'G': 0.0,   'H': -2.0,   'I': -1.0,   'K': -1.0,   'L': -2.0,   'M': -1.0,   'N': -1.0,   'P': -1.0,   'Q': -1.0,   'R': -2.0,   'S': 1.0,   'T': 0.0,   'V': 0.0,   'W': -3.0,   'Y': -2.0},\n",
      "    'C': {'A': -1.0,   'C': 13.0,   'D': -4.0,   'E': -3.0,   'F': -2.0,   'G': -3.0,   'H': -3.0,   'I': -2.0,   'K': -3.0,   'L': -2.0,   'M': -2.0,   'N': -2.0,   'P': -4.0,   'Q': -3.0,   'R': -4.0,   'S': -1.0,   'T': -1.0,   'V': -1.0,   'W': -5.0,   'Y': -3.0},\n",
      "    'D': {'A': -2.0,   'C': -4.0,   'D': 8.0,   'E': 2.0,   'F': -5.0,   'G': -1.0,   'H': -1.0,   'I': -4.0,   'K': -1.0,   'L': -4.0,   'M': -4.0,   'N': 2.0,   'P': -1.0,   'Q': 0.0,   'R': -2.0,   'S': 0.0,   'T': -1.0,   'V': -4.0,   'W': -5.0,   'Y': -3.0},\n",
      "    'E': {'A': -1.0,   'C': -3.0,   'D': 2.0,   'E': 6.0,   'F': -3.0,   'G': -3.0,   'H': 0.0,   'I': -4.0,   'K': 1.0,   'L': -3.0,   'M': -2.0,   'N': 0.0,   'P': -1.0,   'Q': 2.0,   'R': 0.0,   'S': -1.0,   'T': -1.0,   'V': -3.0,   'W': -3.0,   'Y': -2.0},\n",
      "    'F': {'A': -3.0,   'C': -2.0,   'D': -5.0,   'E': -3.0,   'F': 8.0,   'G': -4.0,   'H': -1.0,   'I': 0.0,   'K': -4.0,   'L': 1.0,   'M': 0.0,   'N': -4.0,   'P': -4.0,   'Q': -4.0,   'R': -3.0,   'S': -3.0,   'T': -2.0,   'V': -1.0,   'W': 1.0,   'Y': 4.0},\n",
      "    'G': {'A': 0.0,   'C': -3.0,   'D': -1.0,   'E': -3.0,   'F': -4.0,   'G': 8.0,   'H': -2.0,   'I': -4.0,   'K': -2.0,   'L': -4.0,   'M': -3.0,   'N': 0.0,   'P': -2.0,   'Q': -2.0,   'R': -3.0,   'S': 0.0,   'T': -2.0,   'V': -4.0,   'W': -3.0,   'Y': -3.0},\n",
      "    'H': {'A': -2.0,   'C': -3.0,   'D': -1.0,   'E': 0.0,   'F': -1.0,   'G': -2.0,   'H': 10.0,   'I': -4.0,   'K': 0.0,   'L': -3.0,   'M': -1.0,   'N': 1.0,   'P': -2.0,   'Q': 1.0,   'R': 0.0,   'S': -1.0,   'T': -2.0,   'V': -4.0,   'W': -3.0,   'Y': 2.0},\n",
      "    'I': {'A': -1.0,   'C': -2.0,   'D': -4.0,   'E': -4.0,   'F': 0.0,   'G': -4.0,   'H': -4.0,   'I': 5.0,   'K': -3.0,   'L': 2.0,   'M': 2.0,   'N': -3.0,   'P': -3.0,   'Q': -3.0,   'R': -4.0,   'S': -3.0,   'T': -1.0,   'V': 4.0,   'W': -3.0,   'Y': -1.0},\n",
      "    'K': {'A': -1.0,   'C': -3.0,   'D': -1.0,   'E': 1.0,   'F': -4.0,   'G': -2.0,   'H': 0.0,   'I': -3.0,   'K': 6.0,   'L': -3.0,   'M': -2.0,   'N': 0.0,   'P': -1.0,   'Q': 2.0,   'R': 3.0,   'S': 0.0,   'T': -1.0,   'V': -3.0,   'W': -3.0,   'Y': -2.0},\n",
      "    'L': {'A': -2.0,   'C': -2.0,   'D': -4.0,   'E': -3.0,   'F': 1.0,   'G': -4.0,   'H': -3.0,   'I': 2.0,   'K': -3.0,   'L': 5.0,   'M': 3.0,   'N': -4.0,   'P': -4.0,   'Q': -2.0,   'R': -3.0,   'S': -3.0,   'T': -1.0,   'V': 1.0,   'W': -2.0,   'Y': -1.0},\n",
      "    'M': {'A': -1.0,   'C': -2.0,   'D': -4.0,   'E': -2.0,   'F': 0.0,   'G': -3.0,   'H': -1.0,   'I': 2.0,   'K': -2.0,   'L': 3.0,   'M': 7.0,   'N': -2.0,   'P': -3.0,   'Q': 0.0,   'R': -2.0,   'S': -2.0,   'T': -1.0,   'V': 1.0,   'W': -1.0,   'Y': 0.0},\n",
      "    'N': {'A': -1.0,   'C': -2.0,   'D': 2.0,   'E': 0.0,   'F': -4.0,   'G': 0.0,   'H': 1.0,   'I': -3.0,   'K': 0.0,   'L': -4.0,   'M': -2.0,   'N': 7.0,   'P': -2.0,   'Q': 0.0,   'R': -1.0,   'S': 1.0,   'T': 0.0,   'V': -3.0,   'W': -4.0,   'Y': -2.0},\n",
      "    'P': {'A': -1.0,   'C': -4.0,   'D': -1.0,   'E': -1.0,   'F': -4.0,   'G': -2.0,   'H': -2.0,   'I': -3.0,   'K': -1.0,   'L': -4.0,   'M': -3.0,   'N': -2.0,   'P': 10.0,   'Q': -1.0,   'R': -3.0,   'S': -1.0,   'T': -1.0,   'V': -3.0,   'W': -4.0,   'Y': -3.0},\n",
      "    'Q': {'A': -1.0,   'C': -3.0,   'D': 0.0,   'E': 2.0,   'F': -4.0,   'G': -2.0,   'H': 1.0,   'I': -3.0,   'K': 2.0,   'L': -2.0,   'M': 0.0,   'N': 0.0,   'P': -1.0,   'Q': 7.0,   'R': 1.0,   'S': 0.0,   'T': -1.0,   'V': -3.0,   'W': -1.0,   'Y': -1.0},\n",
      "    'R': {'A': -2.0,   'C': -4.0,   'D': -2.0,   'E': 0.0,   'F': -3.0,   'G': -3.0,   'H': 0.0,   'I': -4.0,   'K': 3.0,   'L': -3.0,   'M': -2.0,   'N': -1.0,   'P': -3.0,   'Q': 1.0,   'R': 7.0,   'S': -1.0,   'T': -1.0,   'V': -3.0,   'W': -3.0,   'Y': -1.0},\n",
      "    'S': {'A': 1.0,   'C': -1.0,   'D': 0.0,   'E': -1.0,   'F': -3.0,   'G': 0.0,   'H': -1.0,   'I': -3.0,   'K': 0.0,   'L': -3.0,   'M': -2.0,   'N': 1.0,   'P': -1.0,   'Q': 0.0,   'R': -1.0,   'S': 5.0,   'T': 2.0,   'V': -2.0,   'W': -4.0,   'Y': -2.0},\n",
      "    'T': {'A': 0.0,   'C': -1.0,   'D': -1.0,   'E': -1.0,   'F': -2.0,   'G': -2.0,   'H': -2.0,   'I': -1.0,   'K': -1.0,   'L': -1.0,   'M': -1.0,   'N': 0.0,   'P': -1.0,   'Q': -1.0,   'R': -1.0,   'S': 2.0,   'T': 5.0,   'V': 0.0,   'W': -3.0,   'Y': -2.0},\n",
      "    'V': {'A': 0.0,   'C': -1.0,   'D': -4.0,   'E': -3.0,   'F': -1.0,   'G': -4.0,   'H': -4.0,   'I': 4.0,   'K': -3.0,   'L': 1.0,   'M': 1.0,   'N': -3.0,   'P': -3.0,   'Q': -3.0,   'R': -3.0,   'S': -2.0,   'T': 0.0,   'V': 5.0,   'W': -3.0,   'Y': -1.0},\n",
      "    'W': {'A': -3.0,   'C': -5.0,   'D': -5.0,   'E': -3.0,   'F': 1.0,   'G': -3.0,   'H': -3.0,   'I': -3.0,   'K': -3.0,   'L': -2.0,   'M': -1.0,   'N': -4.0,   'P': -4.0,   'Q': -1.0,   'R': -3.0,   'S': -4.0,   'T': -3.0,   'V': -3.0,   'W': 15.0,   'Y': 2.0},\n",
      "    'Y': {'A': -2.0,   'C': -3.0,   'D': -3.0,   'E': -2.0,   'F': 4.0,   'G': -3.0,   'H': 2.0,   'I': -1.0,   'K': -2.0,   'L': -1.0,   'M': 0.0,   'N': -2.0,   'P': -3.0,   'Q': -1.0,   'R': -1.0,   'S': -2.0,   'T': -2.0,   'V': -1.0,   'W': 2.0,   'Y': 8.0}}\n",
      "    \n",
      "def blosum50_scorer(x,y):\n",
      "    return blosum50[x][y]\n",
      "\n",
      "# Define a function to compute a score matrix, given two sequences and a score matrix.\n",
      "\n",
      "def generate_score_matrix(seq1,seq2,substitution_matrix):\n",
      "    # Initialize a matrix to use for storing the scores\n",
      "    score_matrix = []\n",
      "    # Iterate over the amino acids in sequence two (which will correspond\n",
      "    # to the vertical sequence in the matrix)\n",
      "    for aa2 in seq2:\n",
      "        # Initialize the current row of the matrix\n",
      "        current_row = []\n",
      "        # Iterate over the amino acids in sequence one (which will\n",
      "        # correspond to the horizontal sequence in the matrix)\n",
      "        for aa1 in seq1:\n",
      "            # score as 1 if the bases are equal and 0 if they're not\n",
      "            current_row.append(blosum50[aa1][aa2])\n",
      "        # append the current row to the matrix\n",
      "        score_matrix.append(current_row)\n",
      "    return score_matrix\n",
      "\n",
      "# Define a function to compute the Needleman-Wunsch matrix and the traceback matrix.\n",
      "\n",
      "def generate_nw_and_traceback_matrices(seq1,seq2,gap_penalty,substitution_matrix):\n",
      "\n",
      "    # Initialize a matrix to use for scoring the alignment and for tracing\n",
      "    # back the best alignment\n",
      "    nw_matrix = [[-gap_penalty * i for i in range(0,len(seq1)+1)]]\n",
      "    traceback_matrix = [[None] + ['-' for i in range(0,len(seq1))]]\n",
      "    # Iterate over the amino acids in sequence two (which will correspond\n",
      "    # to the vertical sequence in the matrix)\n",
      "    # Note that i corresponds to column numbers, as in the 'Biological Sequence\n",
      "    # Analysis' example from class\n",
      "    for i,aa2 in zip(range(1,len(seq2)+1),seq2):\n",
      "        # Initialize the current row of the matrix\n",
      "        current_row = [i * -gap_penalty]\n",
      "        current_traceback_matrix_row = ['|']\n",
      "        # Iterate over the amino acids in sequence one (which will\n",
      "        # correspond to the horizontal sequence in the matrix)\n",
      "        # Note that j corresponds to row numbers, as in the 'Biological Sequence\n",
      "        # Analysis' example from class\n",
      "        for j,aa1 in zip(range(1,len(seq1)+1),seq1):\n",
      "            substitution_score = substitution_matrix[aa1][aa2]\n",
      "            diag_score = (nw_matrix[i-1][j-1] + substitution_score,'\\\\')\n",
      "            up_score = (nw_matrix[i-1][j] - gap_penalty,'|')\n",
      "            left_score = (current_row[-1] - gap_penalty,'-')\n",
      "            best_score = max(diag_score,up_score,left_score)\n",
      "            current_row.append(best_score[0])\n",
      "            current_traceback_matrix_row.append(best_score[1])\n",
      "        # append the current row to the matrix\n",
      "        nw_matrix.append(current_row)\n",
      "        traceback_matrix.append(current_traceback_matrix_row)\n",
      "    return nw_matrix, traceback_matrix\n",
      "\n",
      "# Define a function to perform the traceback.\n",
      "\n",
      "def nw_traceback(traceback_matrix,nw_matrix,seq1,seq2,gap_character='-'):\n",
      "\n",
      "    aligned_seq1 = []\n",
      "    aligned_seq2 = []\n",
      "\n",
      "    current_row = len(traceback_matrix) - 1\n",
      "    current_col = len(traceback_matrix[0]) - 1\n",
      "\n",
      "    best_score = nw_matrix[current_row][current_col]\n",
      "\n",
      "    while True:\n",
      "        current_value = traceback_matrix[current_row][current_col]\n",
      "\n",
      "        if current_value == '\\\\':\n",
      "            aligned_seq1.append(seq1[current_col-1])\n",
      "            aligned_seq2.append(seq2[current_row-1])\n",
      "            current_row -= 1\n",
      "            current_col -= 1\n",
      "        elif current_value == '|':\n",
      "            aligned_seq1.append('-')\n",
      "            aligned_seq2.append(seq2[current_row-1])\n",
      "            current_row -= 1\n",
      "        elif current_value == '-':\n",
      "            aligned_seq1.append(seq1[current_col-1])\n",
      "            aligned_seq2.append('-')\n",
      "            current_col -= 1\n",
      "        elif current_value == None:\n",
      "            break\n",
      "        else:\n",
      "            raise ValueError, \"Invalid value in traceback matrix: %s\" % current_value\n",
      "\n",
      "    return ''.join(aligned_seq1[::-1]), ''.join(aligned_seq2[::-1]), best_score\n",
      "\n",
      "# Define some functions to conveniently format the score matrix and the dynamic programming (in our case the Needleman-Wunsch) matrix for viewing.\n",
      "\n",
      "def format_score_matrix(seq1,seq2,score_matrix,title):\n",
      "    # define a format string that will be used to format each line -\n",
      "    # since seq1 is the 'horizontal' sequence in our matrix we'll\n",
      "    # have len(seq1) + 1 entries on each line to print the scores and\n",
      "    # seq2 base associated with each row\n",
      "    line_format = \"%6s\" * (len(seq1) + 1)\n",
      "\n",
      "    print \"\\n%s\" % title\n",
      "\n",
      "    # print seq1 (start the line with an empty string)\n",
      "    print line_format % tuple([' '] + map(str,list(seq1)))\n",
      "\n",
      "    # iterate over the rows and print each (starting with the\n",
      "    # corresponding base in sequence2)\n",
      "    for row, base in zip(score_matrix,seq2):\n",
      "        print line_format % tuple([base] + map(str,row))\n",
      "\n",
      "def format_dynamic_programming_matrix(seq1,seq2,matrix,title):\n",
      "    print \"\\n%s\" % title\n",
      "\n",
      "    line_format = \"%6s\" * (len(seq1) + 2)\n",
      "    # print seq1 (start the line with two empty strings)\n",
      "    print line_format % tuple([' ',' '] + map(str,list(seq1)))\n",
      "\n",
      "    # iterate over the rows and print each (starting with the\n",
      "    # corresponding base in sequence2)\n",
      "    for row, base in zip(matrix,' ' + seq2):\n",
      "        print line_format % tuple([base] + map(str,row))\n",
      "        \n",
      " # Main program\n",
      " \n",
      "\n",
      "seq1 = \"HEAGAWGHEE\"\n",
      "seq2 = \"PAWHEAE\"\n",
      "\n",
      "score_matrix = generate_score_matrix(seq1,seq2,blosum50)\n",
      "\n",
      "format_score_matrix(seq1, seq2, score_matrix, title=\"Score matrix (based on BLOSUM50)\")\n",
      "\n",
      "nw_matrix, traceback_matrix = generate_nw_and_traceback_matrices(seq1,seq2, 8, blosum50)\n",
      "\n",
      "aligned_seq1, aligned_seq2, score = nw_traceback(traceback_matrix,nw_matrix,seq1,seq2)\n",
      "\n",
      "format_dynamic_programming_matrix(seq1, seq2, nw_matrix, title=\"Global dynamic programming matrix\")\n",
      "\n",
      "format_dynamic_programming_matrix(seq1, seq2, traceback_matrix, title=\"Traceback matrix\")\n",
      "\n",
      "\n",
      "print \"\\nAlignment:\"\n",
      "\n",
      "print aligned_seq1\n",
      "print aligned_seq2\n",
      "print '\\nAlignment score:'\n",
      "print score"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Score matrix (based on BLOSUM50)\n",
        "           H     E     A     G     A     W     G     H     E     E\n",
        "     P  -2.0  -1.0  -1.0  -2.0  -1.0  -4.0  -2.0  -2.0  -1.0  -1.0\n",
        "     A  -2.0  -1.0   5.0   0.0   5.0  -3.0   0.0  -2.0  -1.0  -1.0\n",
        "     W  -3.0  -3.0  -3.0  -3.0  -3.0  15.0  -3.0  -3.0  -3.0  -3.0\n",
        "     H  10.0   0.0  -2.0  -2.0  -2.0  -3.0  -2.0  10.0   0.0   0.0\n",
        "     E   0.0   6.0  -1.0  -3.0  -1.0  -3.0  -3.0   0.0   6.0   6.0\n",
        "     A  -2.0  -1.0   5.0   0.0   5.0  -3.0   0.0  -2.0  -1.0  -1.0\n",
        "     E   0.0   6.0  -1.0  -3.0  -1.0  -3.0  -3.0   0.0   6.0   6.0\n",
        "\n",
        "Global dynamic programming matrix\n",
        "                 H     E     A     G     A     W     G     H     E     E\n",
        "           0    -8   -16   -24   -32   -40   -48   -56   -64   -72   -80\n",
        "     P    -8  -2.0  -9.0 -17.0 -25.0 -33.0 -41.0 -49.0 -57.0 -65.0 -73.0\n",
        "     A   -16 -10.0  -3.0  -4.0 -12.0 -20.0 -28.0 -36.0 -44.0 -52.0 -60.0\n",
        "     W   -24 -18.0 -11.0  -6.0  -7.0 -15.0  -5.0 -13.0 -21.0 -29.0 -37.0\n",
        "     H   -32 -14.0 -18.0 -13.0  -8.0  -9.0 -13.0  -7.0  -3.0 -11.0 -19.0\n",
        "     E   -40 -22.0  -8.0 -16.0 -16.0  -9.0 -12.0 -15.0  -7.0   3.0  -5.0\n",
        "     A   -48 -30.0 -16.0  -3.0 -11.0 -11.0 -12.0 -12.0 -15.0  -5.0   2.0\n",
        "     E   -56 -38.0 -24.0 -11.0  -6.0 -12.0 -14.0 -15.0 -12.0  -9.0   1.0\n",
        "\n",
        "Traceback matrix\n",
        "                 H     E     A     G     A     W     G     H     E     E\n",
        "        None     -     -     -     -     -     -     -     -     -     -\n",
        "     P     |     \\     \\     \\     -     \\     -     -     -     \\     \\\n",
        "     A     |     |     \\     \\     -     \\     -     -     -     -     -\n",
        "     W     |     |     |     \\     \\     \\     \\     -     -     -     -\n",
        "     H     |     \\     \\     \\     \\     \\     |     \\     \\     -     -\n",
        "     E     |     |     \\     -     |     \\     \\     |     \\     \\     \\\n",
        "     A     |     |     |     \\     -     \\     \\     \\     |     |     \\\n",
        "     E     |     |     |     |     \\     \\     \\     \\     \\     \\     \\\n",
        "\n",
        "Alignment:\n",
        "HEAGAWGHE-E\n",
        "--P-AW-HEAE\n",
        "\n",
        "Alignment score:\n",
        "1.0\n"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Software for sequence alignment\n",
      "\n",
      "There are many tools available for sequence alignment.  The common tools are hosting at the European Bioinformatics Institute.\n",
      "\n",
      "* Pairwise sequence alignment - http://www.ebi.ac.uk/Tools/psa/\n",
      "* Multiple sequence alignment - http://www.ebi.ac.uk/Tools/msa/"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Exercises\n",
      "\n",
      "1. Using the Dynamic Progamming example in Eric Rouchka's simple DP tutorial as a guide by hand fill out the matrices for the sequences TGTAAT and TCTATTTC.  Turn this is as a word document or pdf\n",
      "\n",
      "2. Modify the above python program to implement the PAM250 scoring matrix - http://www.icp.ucl.ac.be/~opperd/private/pam250.html instead of the Blosum50 matrix.\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "* Next - <a href=\"http://nbviewer.ipython.org/urls/raw.github.com/jeffreyblanchard/AdvGen/master/Session9.ipynb\">Session 9 : Word Tuples and Blast </a>\n",
      "* Previous - <a href=\"http://nbviewer.ipython.org/urls/raw.github.com/jeffreyblanchard/AdvGen/master/Session7.ipynb\">Session 7 : Functions and Modules</a> "
     ]
    }
   ],
   "metadata": {}
  }
 ]
}